framebuf: Fix crash in scroll().

jepler opened the pull request:
### Summary

Fuzzing found a crash in framebuf scroll. If mp_int_t is wider than int, then the tests such as `xend < 0` can fail even when the amount of scrolling requested is out of range. This resulted in a segmentation fault when attempting an out-of-bounds access to the framebuffer.

### Testing

I added a new test for the problem. The try/except is for systems with 32-bit mp_int_t (they throw an exception with an argument this large)

### Trade-offs and Alternatives

Widening the other locals to mp_int_t makes no difference on platforms where it's the same width as int, which is the case for the most constrained devices. However, it might expand nanbox builds a bit in code size.

jepler committed:
framebuf: Fix crash in scroll().

If mp_int_t is wider than int, then the tests such as `xend < 0`
can fail even when the amount of scrolling requested is out of
range. This resulted in a segmentation fault when attempting an
out-of-bounds access to the framebuffer.

Signed-off-by: Jeff Epler <jepler@gmail.com>

dpgeorge commented:
Would be good to get the code size increase for unix x64 down to zero if possible.

Maybe worth thinking a bit about the types needed here:
- `dx` and `dy` can probably stay as `int`
- `sx` and `y` can probably be `unsigned int`, they are guaranteed to be non-negative
- maybe `x` (in the for loop) can be `unsigned int` as well, at least that matches the argument type for `getpixel` and `setpixel`.

jepler commented:
Updated. The best variant I could find locally was +3 bytes in unix standard build on x86_64. Let's see what CI says.

jepler committed:
modframebuf: Save code size in setpixel.

There's a slight code size increase paid for by using
setpixel_checked for the last pixel of a line, instead of
repeating the checks inline.

Signed-off-by: Jeff Epler <jepler@gmail.com>

jepler commented:
> But, maybe just changing the call at the end of the function to use `setpixel_checked()` would be enough to gain most of the code savings, and then we don't need to think about/measure the speed impact of the other calls?

OK, trying that in the newest build.

dpgeorge commented:
This is looking good now, a nice little win overall in terms of code size.
